examples/get-media-details/main.py

```diff
@@ -0,0 +1,73 @@
+import json
+
+from utils.utils import PublicAPIClient, get_standard_arguments
+
+
+def main():
+    args = get_standard_arguments(
+        [
+            (["media_id"], {"type": int, "help": "the media to get details about"}),
+        ]
+    )
+    media_id = args.media_id
+
+    public_api_client = PublicAPIClient(args.subdomain)
+    media = get_media_details(public_api_client, media_id)
+    captions = get_media_details(public_api_client, media_id, "caption_files")
+    courses = get_media_details(public_api_client, media_id, "courses")
+    tags = get_media_details(public_api_client, media_id, "tags")
+    users = get_media_details(public_api_client, media_id, "users", "user_permissions")
+
+    response = get_subdict(
+        media,
+        [
+            "id",
+            "title",
+            "description",
+            "duration",
+            "size",
+            "created_at",
+            "owner",
+            "collection",
+        ],
+    )
+    response.update(
+        {
+            "captions": captions,
+            "courses": courses,
+            "tags": tags,
+            "users": users,
+        }
+    )
+
+    print(
+        json.dumps(
+            response,
+            indent=4,
+            sort_keys=True,
+        )
+    )
+
+
+def get_media_details(
+    public_api_client, media_id, detail_url=None, detail_attribute=None
+):
+    if detail_url:
```

Comment:
@kasra-r77 imho what you really want instead is to turn this module into a class. That would make sense to me.

Reply:
Nope. I think you better read about variable scoping. 😄 

Variables are local by default, so every time a function is called, you have access to the arguments passed and a clean state in which you can define and assign to local variables. This is true for almost every modern language.

Otherwise you would get "magic" in the form of global variables which means you call a function and you will have no idea what will happen. Maybe it runs like this, maybe it runs like that, depending on what other stuff you did in other places of your code.

In this case: who sets media_id and when and where?

Yes, you can set it in main, but that would be main's local variable.

Or you make it global, and then you get the problem I mentioned above.

One valid use case for global variables is constant configuration: you define it at the top of your module (use ALL_CAPS to indicate constantness) and use it everywhere. But then it's pretty clear that's it's const and it's truly global (at least within your module), so it doesn't matter when and where you are calling functions, they will always be available and the same.

In Rails controllers we sometimes use instance variables which *look* like globals, but there's actually a controller object created for every request and the variable belongs to that. In that case it looks magical, because Rails hides it from you that it creates an object for every request.

So if you wanna keep it clear how that data is flowing in your code, you better explicitly set and pass variables. As the [Zen of Python](https://en.wikipedia.org/wiki/Zen_of_Python) says "Explicit is better than implicit."

One way to do it here is to create a detail getter object with subdomain and media_id passed to it and convert get_media_details to a method of that object. Then it's clear how your data moves, but you also don't have to pass them multiple times.

Reply:
Sorry, the "Nope." was a response to Kasra, not Zoli. 😄 

I just pushed the "classified" version.